### PR TITLE
luajit: Expose build options, enable JIT debug module

### PR DESCRIPTION
**Pull request by Shados**
###### Motivation for this change
This change exposes various build options from luajit's makefile, many of which are useful when debugging Lua scripts. I've also enabled installing the JIT debug module files by default (used for `-jbc`, `-jv`, and `-jdump` CLI options).

###### Things done
- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] ~~Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))~~
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] ~~Ensured that relevant documentation is up to date~~
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

@teto

### CODE REVIEW

**Review by lblasc — APPROVED** [review]

**Comment by teto** [comment]
sorry for delay, I don't think we have to map every option in nix. Like not sure `enableJIT` would be common enough to warrant its own flag ?
Valgrind and Gdb could go under the same flag `Debug`. One cqn override XCFLAGS.

**Comment by Shados** [comment]
> sorry for delay, I don't think we have to map every option in nix. Like not sure `enableJIT` would be common enough to warrant its own flag ?

There's no real reason *not* to expose all the options, though. They're not likely to change. As for `enableJIT`: running without the JIT compiler makes some debugging scenarios easier.

> Valgrind and Gdb could go under the same flag `Debug`.

Come to think of it, pretty sure the valgrind option won't work as-is, IIRC it needs to valgrind header files to build. I'll have a poke at that later, but it should definitely be separate: the GDB JIT option causes a performance impact (even without GDB hooked up),  and is mainly useful for debugging LuaJIT itself (whereas the valgrind option is potentially useful for debugging leaks in Lua scripts in general, at least insofar as ones that use C libraries).

File: pkgs/development/interpreters/luajit/default.nix [hunk 1]
```diff
@@ -22,27 +44,34 @@ stdenv.mkDerivation rec {
...
+      # CCDEBUG must be non-empty or everything will be stripped, -g being
+      # passed by nixpkgs CC wrapper is insufficient on its own
+      substituteInPlace src/Makefile --replace "#CCDEBUG= -g" "CCDEBUG= -g"
+    fi
   '';
 
   configurePhase = false;
 
+  buildFlags = [
+    "amalg" # Build highly optimized version
```
Comment (teto) [inline pkgs/development/interpreters/luajit/default.nix]: is there a downside to this ?

**Comment by teto** [comment]
> There's no real reason not to expose all the options, though. They're not likely to change.

It makes the nix derivation harder to read for a usecase that may benefit 1% of the use cases. But you convinced me anyway and as you've been pretty helpful in the past, I know you'll make good use of these options. Let's fix the valgrind option and merge.

File: pkgs/development/interpreters/luajit/default.nix (hunk 1 above)
Comment (Shados) [reply pkgs/development/interpreters/luajit/default.nix]: To my knowledge, the only downside is that the *build* process with `amalg` enabled consumes more memory. `amalg` is the recommended target for binary distribution.

**Comment by Shados** [comment]
OK, fixed up the valgrind option.

**Pull request merged**